### PR TITLE
[native pos] Add end to end functionality for file based broadcast

### PR DESCRIPTION
**Pull request by singcha**
Add broadcast support for file based broadcast by :

- Adding plan node conversion to use BroadcastWriteNode for broadcast output & BroadcastExchangeSource for broadcast read
- Adding code changes to pass baseBroadcastPath & list of broadcast files

Overall flow after this change :
1. Executor submits plan fragment with broadcast output to native process
2. native process converts the plan fragment by adding BroadcastWriteNode, which writes to file system
3. File details(currently file path only) is collected back and returned to spark driver
4. Driver broadcasts file details in next stage to all executors
5. Executors send plan fragment with remote source as filePaths to native process
6. Native process uses BroadcastExchangeSource to read back broadcast table

Test plan - 
Tested by enabling disabled tests which uses broadcast.

```
== RELEASE NOTES ==

General Changes
* Add support for broadcast join in Presto-on-Spark/Velox execution path
* Added new property `native-execution-broadcast-base-path` which is used to specify base path for temporary storage of broadcast data for presto-on-spark native execution

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 1]
```diff
@@ -643,6 +651,9 @@ void PrestoServer::registerCustomOperators() {
       std::make_unique<facebook::presto::operators::ShuffleWriteTranslator>());
   facebook::velox::exec::Operator::registerOperator(
       std::make_unique<operators::ShuffleReadTranslator>());
+  facebook::velox::exec::Operator::registerOperator(
+      std::make_unique<
+          facebook::presto::operators::BroadcastWriteTranslator>());
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:656]: This is POS-specific, right? Let's split Prestissimo and POS into 2 separate classes to avoid mixing the logic.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 2]
```diff
@@ -213,11 +213,13 @@ class VeloxBatchQueryPlanConverter : public VeloxQueryPlanConverterBase {
   VeloxBatchQueryPlanConverter(
       const std::string& shuffleName,
       std::shared_ptr<std::string>&& serializedShuffleWriteInfo,
+      std::shared_ptr<std::string>&& broadcastBasePath,
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h:216]: Use std::optional<std::string>.

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp [hunk 3]
```diff
@@ -146,10 +146,16 @@ velox::BufferPtr BroadcastFileReader::next() {
 
   auto readFile = fileSystem_->openFileForRead(
       broadcastFileInfos_[readfileIndex_].filePath_);
-  auto buffer = AlignedBuffer::allocate<char>(readFile->size(), pool_, 0);
-  readFile->pread(0, readFile->size(), buffer->asMutable<char>());
   ++readfileIndex_;
   ++numFiles_;
+
+  // Empty broadcast file.
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:152]: Is this a bug fix? Would you extract it into a separate PR and add a test?

File: presto-native-execution/presto_cpp/presto_protocol/README.md [hunk 4]
```diff
@@ -2,10 +2,11 @@
 
 ### Required Dependencies:
 
-Install [PyYAML](https://pyyaml.org/) and [Chevron](https://github.com/noahmorrison/chevron)
+Install [PyYAML](https://pyyaml.org/), [Chevron](https://github.com/noahmorrison/chevron) and [jq](https://github.com/jqlang/jq)
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/presto_protocol/README.md:5]: Would you extract these changes into a separate PR? CC: @majetideepak

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeSimpleQueries.java [hunk 5]
```diff
@@ -116,4 +119,17 @@ public void testJsonFileBasedFunction()
         assertQuery("SELECT json.test_schema.eq(1, linenumber) FROM lineitem", "SELECT 1 = linenumber FROM lineitem");
         assertQuery("SELECT json.test_schema.sum(linenumber) FROM lineitem", "SELECT sum(linenumber) FROM lineitem");
     }
+
+    @Test
+    public void testBroadcastJoin()
```
Comment (mbasmanova) [inline presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeSimpleQueries.java:124]: Is this needed? Existing join tests should be sufficient, no?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark;
+
+public interface BroadcastStorage
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java:16]: Please, add comments.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/LocalBroadcastStorage.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+public class LocalBroadcastStorage
+        implements BroadcastStorage
+{
+    @Override
+    public String getBroadcastDirectoryPath(String queryId)
+    {
+        String directoryPrefix = format("%s_%s", "local_broadcast", queryId);
+        try {
+            return createTempDirectory(directoryPrefix).toAbsolutePath().toString();
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/LocalBroadcastStorage.java:29]: I imagine this won't work on a cluster.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java [hunk 8]
```diff
@@ -26,14 +26,17 @@ public class BatchTaskUpdateRequest
...
     private final TaskUpdateRequest taskUpdateRequest;
     private final Optional<String> shuffleWriteInfo;
+    private final Optional<String> broadcastBasePath;
 
     @JsonCreator
     public BatchTaskUpdateRequest(
             @JsonProperty("taskUpdateRequest") TaskUpdateRequest taskUpdateRequest,
-            @JsonProperty("shuffleWriteInfo") Optional<String> shuffleWriteInfo)
+            @JsonProperty("shuffleWriteInfo") Optional<String> shuffleWriteInfo,
+            @JsonProperty("broadcastBasePath") Optional<String> broadcastBasePath)
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java:35]: Can we re-use shuffleWriteInfo? In case of broadcast, it can store the broadcast base path, no?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * This class is a 1:1 strict API mapping to BroadcastFileInfo in
+ * presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h.
+ * Please refrain changes to this API class. If any changes have to be made to
+ * this class, one should make sure to make corresponding changes in the above
+ * C++ struct and its corresponding serde functionalities.
+ */
+public class BroadcastFileInfo
+{
+    private String filePath;
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java:30]: Do you expect adding other fields? If so, what are these? If not, we probably don't need this class.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * This class is a 1:1 strict API mapping to BroadcastInfo in
+ * presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h.
+ * Please refrain changes to this API class. If any changes have to be made to
+ * this class, one should make sure to make corresponding changes in the above
+ * C++ struct and its corresponding serde functionalities.
+ */
+public class BroadcastInfo
+{
+    private String basePath;
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java:33]: Make these final.

**Comment by mbasmanova** [comment]
@pgupta2 Arjun pointed out that we may need to implement a cache of broadcast data to avoid reading same data multiple times from the same executor in case multiple tasks from the same stage end up on that executor. Here is Java implementation: https://github.com/prestodb/presto/blob/master/presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkBroadcastTableCacheManager.java

File: presto-native-execution/presto_cpp/presto_protocol/README.md (hunk 4 above)
Comment (majetideepak) [reply presto-native-execution/presto_cpp/presto_protocol/README.md:5]: Let's move this to a separate PR. `brew jq` is MacOS specific. We should probably update the setup scripts to include jq.

File: presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeSimpleQueries.java (hunk 5 above)
Comment (singcha) [reply presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeSimpleQueries.java:124]: Existing should be sufficient. Removed.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/LocalBroadcastStorage.java (hunk 7 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/LocalBroadcastStorage.java:29]: Yes, this is local storage meant for testing. Distributed storage specific implementations will be added.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java (hunk 8 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java:35]: I have kept it separate to avoid mixing up logic. Also, there are existing code paths(logic) which depend on this variable being populated : https://github.com/search?q=repo%3Aprestodb%2Fpresto%20shuffleWriteInfo&type=code

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 1 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:671]: Good point. I was following existing shuffle registration. Should I refactor shuffle ones too ? Is it fine to do this refactoring in seperate PR?

File: presto-native-execution/presto_cpp/presto_protocol/README.md (hunk 4 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/presto_protocol/README.md:5]: https://github.com/prestodb/presto/pull/20052

File: presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp (hunk 3 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/operators/BroadcastFactory.cpp:152]: Yes, raised a separate PR with test : https://github.com/prestodb/presto/pull/20053

**Comment by singcha** [comment]
@mbasmanova - Thanks Masha for review. I have incorporated most of the suggestion and commented on some. Could you please review.

**Comment by singcha** [comment]
> @pgupta2 Arjun pointed out that we may need to implement a cache of broadcast data to avoid reading same data multiple times from the same executor in case multiple tasks from the same stage end up on that executor. Here is Java implementation: https://github.com/prestodb/presto/blob/master/presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkBroadcastTableCacheManager.java

Thank you for pointers. I will raise a follow up PR

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 1 above)
Comment (mbasmanova) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:671]: Yes, let's refactor in a separate PR (a prerequisite to this one). Thanks.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 11]
```diff
@@ -309,9 +311,15 @@ void PrestoServer::run() {
...
   facebook::velox::exec::ExchangeSource::registerFactory(
       PrestoExchangeSource::createExchangeSource);
+
+  // Batch partitioned shuffle exchange source.
   facebook::velox::exec::ExchangeSource::registerFactory(
       operators::UnsafeRowExchangeSource::createExchangeSource);
 
+  // Batch broadcast exchange source.
+  facebook::velox::exec::ExchangeSource::registerFactory(
+      operators::BroadcastExchangeSource::createExchangeSource);
```
[the remark this answers is not in the document]
Comment (singcha) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:322]: We will need to add registerCustomOperators too as vitual method and both of these will have pos vs presto specific implementation. Please see this comment - https://github.com/prestodb/presto/pull/19917/files#r1242629653

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h (hunk 2 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h:220]: presto_protocol generated batchUpdaterequest produces this type definition. Should I create optional out of shared ptr here ?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 12]
```diff
@@ -2550,9 +2550,24 @@ velox::core::PlanFragment VeloxBatchQueryPlanConverter::toVeloxQueryPlan(
...
-      "Broadcast shuffle is not supported");
+  if (partitionedOutputNode->isBroadcast()) {
+    VELOX_USER_CHECK_NOT_NULL(
+        broadcastBasePath_, "broadcastBasePath is required");
+    auto broadcastWriteNode = std::make_shared<operators::BroadcastWriteNode>(
+        "broadcast-write",
+        *broadcastBasePath_,
+        core::LocalPartitionNode::gather(
+            "broadcast-write-gather",
+            std::vector<core::PlanNodePtr>{
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2561]: This can be just partitionedOutputNode->sources(), no?

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp [hunk 13]
```diff
@@ -2550,9 +2550,24 @@ velox::core::PlanFragment VeloxBatchQueryPlanConverter::toVeloxQueryPlan(
...
+                partitionedOutputNode->sources()[0]}));
+
+    planFragment.planNode = core::PartitionedOutputNode::broadcast(
+        "partitioned-output",
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2565]: Using these plan nodes prevents us from reporting the stats as the stats are attached to original plan nodes. Consider using the original plan node for at least one of these and maybe figure something out in processOperatorStats in presto-native-execution/presto_cpp/main/PrestoTask.cpp

File: presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark;
+
+/**
+ * BroadcastStorage specifies contract for specifying underlying file systems used for
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java:17]: This comment is too cryptic. What is this interface for? Do we need it? 

CC: @shrinidhijoshi @vermapratyush

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java [hunk 15]
```diff
@@ -457,6 +461,9 @@ protected void setup(Binder binder)
         binder.bind(SpillerStats.class).in(Scopes.SINGLETON);
         binder.bind(StandaloneSpillerFactory.class).to(TempStorageStandaloneSpillerFactory.class).in(Scopes.SINGLETON);
 
+        // Broadcast file storage.
+        binder.bind(BroadcastStorage.class).to(LocalBroadcastStorage.class).in(Scopes.SINGLETON);
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:465]: It seems strange to hard-code LocalBroadcastStorage. Shouldn't this be driven by configuration?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java [hunk 16]
```diff
@@ -26,14 +26,17 @@ public class BatchTaskUpdateRequest
...
     {
         this.taskUpdateRequest = requireNonNull(taskUpdateRequest, "taskUpdateRequest is null");
         this.shuffleWriteInfo = requireNonNull(shuffleWriteInfo, "shuffleWriteInfo is null");
+        this.broadcastBasePath = requireNonNull(broadcastBasePath, "broadcastBasePath is null");
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java:39]: Should we check that only one of broadcastBasePath or shuffleWriteInfo is set? Ideally, we'd use just one property though.

CC: @shrinidhijoshi @vermapratyush

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java [hunk 17]
```diff
@@ -48,12 +51,19 @@ public Optional<String> getShuffleWriteInfo()
...
+    {
+        return broadcastBasePath;
+    }
+
     @Override
     public String toString()
     {
         return toStringHelper(this)
                 .add("taskUpdateRequest", taskUpdateRequest)
                 .add("shuffleWriteInfo", shuffleWriteInfo)
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java:65]: Would be nice to include this only if set. Same for broadcastBasePath.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+ * This class is a 1:1 strict API mapping to BroadcastFileInfo in
+ * presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h.
+ * Please refrain changes to this API class. If any changes have to be made to
+ * this class, one should make sure to make corresponding changes in the above
+ * C++ struct and its corresponding serde functionalities.
+ */
+public class BroadcastFileInfo
+{
+    private final String filePath;
+    // TODO: Add additional stats including checksum, num rows, size.
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java:31]: Any particular reason not to include number of rows and size now?

**Review by mbasmanova — COMMENTED** [review]
Some further comments.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java [hunk 19]
```diff
@@ -372,11 +401,33 @@ private List<TaskSource> getTaskSources(
...
+                    List<BroadcastFileInfo> fileInfos =
+                            ((List<?>) broadcast.value()).stream()
+                                    .map(PrestoSparkSerializedPage.class::cast)
+                                    .map(prestoSparkSerializedPage -> PrestoSparkUtils.toSerializedPage(prestoSparkSerializedPage))
+                                    .map(serializedPage -> pagesSerde.deserialize(serializedPage))
+                                    // Extract filePath.
+                                    .flatMap(page -> IntStream.range(0, page.getPositionCount())
+                                            .mapToObj(position -> VarcharType.VARCHAR.getObjectValue(null, page.getBlock(0), position)))
+                                    .map(String.class::cast)
+                                    .map(filePath -> new BroadcastFileInfo(filePath))
```
Comment (shrinidhijoshi) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java:416]: A split usually represents a part of a file or the whole file.  Instead of supplying all file metadata as part of `broadcastInfo` as parameter to a single split, should we be creating a `ScheduledSplit` per file,  ? 
We could even go as far as setting the path to be the actual `ws` path in the `Location` field of the Split like so ?

```Java
...
.map(filePath -> new ScheduledSplit(
    nextSplitId.getAndIncrement(), 
    remoteSource.getId(), 
    new Split(
        REMOTE_CONNECTOR_ID, 
        new RemoteTransactionHandle(), 
        new RemoteSplit(new Location(broadcastDirectory + filePath)))) // <<-- FILE PATH HERE ?
.collect(toImmutableList());
```

I can see that we are just following the `shuffleInfo` pattern of stuffing everything into a metadata object, but that is probably not very well designed.

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h [hunk 20]
```diff
@@ -217,11 +217,13 @@ class VeloxBatchQueryPlanConverter : public VeloxQueryPlanConverterBase {
...
       velox::core::QueryCtx* queryCtx,
       velox::memory::MemoryPool* pool)
       : VeloxQueryPlanConverterBase(queryCtx, pool),
         shuffleName_(shuffleName),
-        serializedShuffleWriteInfo_(std::move(serializedShuffleWriteInfo)) {}
+        serializedShuffleWriteInfo_(std::move(serializedShuffleWriteInfo)),
+        broadcastBasePath_(std::move(broadcastBasePath)) {}
```
Comment (shrinidhijoshi) [inline presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h:226]: Would be good to eliminate this  altogether once if we follow the Split generation approach suggested in my other comments

File: presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java (hunk 14 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java:17]: Yeah. I believe something like this will be needed to generate the correct file paths during broadcast write. This will handle 
1. generating a basepath for that shuffle stage `QueryId.StageId.StageExecutionId` 
2. add a prefix of local path `/tmp/shuffle_data` ? or a prefix of the blob storage `ws://pos_prod/` depending on what's being used

Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java:17]: If my understanding is correct @singcha could you please update the comments to be descriptive like above ?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java (hunk 15 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:468]: @singcha  Maybe we can add an `AbstractBroadcastStorage` which throws an `UnsupportedOperation` exception to indicate that the users need to implement their own impl before using this codepath. Also assuming we will bind to `LocalBroadcastStorage` for test modules

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * This class is a 1:1 strict API mapping to BroadcastInfo in
+ * presto-native-execution/presto_cpp/main/operators/BroadcastFactory.h.
+ * Please refrain changes to this API class. If any changes have to be made to
+ * this class, one should make sure to make corresponding changes in the above
+ * C++ struct and its corresponding serde functionalities.
+ */
+public class BroadcastInfo
```
Comment (shrinidhijoshi) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java:32]: Assuming this class will not be needed if the split generation change I suggested in other comments is acceptable.

**Review by shrinidhijoshi — COMMENTED** [review]
Thanks for working on this @singcha ! Left a few comments the main one being 

If we modify the Split generation logic for broadcast read you could reduce some of the metadata related code/classes.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java (hunk 16 above)
Comment (shrinidhijoshi) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java]: I suggested an approach where this field might not be needed. But yes, if have to retain this field then we should do a sanity check

File: presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java (hunk 14 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/BroadcastStorage.java]: Removed this interface and derived class. This is now specified with config

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java (hunk 15 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java]: Not applicable anymore

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java [hunk 22]
```diff
@@ -285,6 +287,8 @@ protected void setup(Binder binder)
         jsonCodecBinder(binder).bindJsonCodec(PrestoSparkLocalShuffleReadInfo.class);
         jsonCodecBinder(binder).bindJsonCodec(PrestoSparkLocalShuffleWriteInfo.class);
         jsonCodecBinder(binder).bindJsonCodec(BatchTaskUpdateRequest.class);
+        jsonCodecBinder(binder).bindJsonCodec(BroadcastFileInfo.class);
```
Comment (vermapratyush) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:290]: As suggested in another PR: https://github.com/prestodb/presto/pull/19820#discussion_r1259581771

Please see if you can merge all these codec into 1 collection and inject using mutli-bind.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java (hunk 19 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java:417]: +1
Try to avoid branching in code as much as possible. `ScheduledSplit` should be able to hide away the split complexity or broadcast path or non-broadcast path.

**Review by vermapratyush — CHANGES_REQUESTED** [review]

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp (hunk 12 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp]: Yes, updated

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java (hunk 18 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastFileInfo.java:31]: Its not added on native side. I was planning on adding it end to end in incremental PR, where we start enforcing size checks

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/BatchTaskUpdateRequest.java [hunk 23]
```diff
@@ -48,12 +51,19 @@ public Optional<String> getShuffleWriteInfo()
         return shuffleWriteInfo;
     }
 
+    @JsonProperty
```
[the remark this answers is not in the document]
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/BatchTaskUpdateRequest.java:55]: Split generation is not related as this info is used during write path

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java (hunk 19 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/task/PrestoSparkNativeTaskExecutorFactory.java:416]: Good call. 
I have updated PR

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java (hunk 21 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BroadcastInfo.java]: Correct, removed

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h (hunk 20 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.h:226]: Split generation is not related as this info is used during write path

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java (hunk 16 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/BatchTaskUpdateRequest.java]: Thanks, added check to validate only one is present

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java (hunk 22 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:289]: Can you please clarify what multi-bind means ?

Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:289]: [multi-bind](https://google.github.io/guice/api-docs/3.0/javadoc/com/google/inject/multibindings/Multibinder.html). You can bind collections like List, Map, Set.

**Review by shrinidhijoshi — COMMENTED** [review]
One last comment on the property name. 

Also, consider below changes on the commit message
1. Wrap the text rather than a single long line
2. Add more details about changes the commit introduces.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (singcha) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:295]: Added TODO for splitting presto and pos server

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 1 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:683]: Added TODO for splitting presto and pos server

File: presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp (hunk 13 above)
Comment (singcha) [reply presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp:2714]: Added TODO reusing plan node and aggregating stats

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java (hunk 22 above)
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkModule.java:289]: We can address this refactoring in followup PR

**Comment by singcha** [comment]
> One last comment on the property name.
> 
> Also, consider below changes on the commit message
> 
> 1. Wrap the text rather than a single long line
> 2. Add more details about changes the commit introduces.

Done, Thanks for pointing it out

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java [hunk 24]
```diff
@@ -477,4 +478,17 @@ public PrestoSparkConfig setAdaptiveJoinSideSwitchingEnabled(boolean adaptiveJoi
         this.adaptiveJoinSideSwitchingEnabled = adaptiveJoinSideSwitchingEnabled;
         return this;
     }
+
+    public String getNativeExecutionBroadcastFilePath()
+    {
+        return nativeExecutionBroadcastFilePath;
+    }
+
+    @Config("native-execution-broadcast-file-path")
```
[the remark this answers is not in the document]
Comment (singcha) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java]: Good call, renamed to native-execution-broadcast-base-path to keep it consistent with native naming

**Comment by singcha** [comment]
@mbasmanova @vermapratyush @shrinidhijoshi - Thank you for the review. I have addressed these and below items will be addressed in followup PR
1. Refactoring - Using multi-bind for JSON codecs
2. Refactoring - Splitting Presto & POS server
3. Re-using and aggregating plan node for stats generation

Can you please review.

**Comment by shrinidhijoshi** [comment]
@singcha Please update the release notes section with the new property added.

**Comment by singcha** [comment]
> @singcha Please update the release notes section with the new property added.

Added, thanks!

**Review by shrinidhijoshi — APPROVED** [review]
LGTM @singcha . Thanks for working on this and addressing all the feedback.

**Review by vermapratyush — APPROVED** [review]
Thanks for addressing the feedback @singcha

**Review by mbasmanova — APPROVED** [review]
Thanks.
